jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DelayedHandler.java

```diff
@@ -163,75 +197,146 @@ protected Callback getCallback()
         }
 
         protected void process()
+        {
+            process(getRequest(), getResponse(), getCallback());
+        }
+
+        protected boolean process(Request request, Response response, Callback callback)
         {
             try
             {
-                if (!getHandler().handle(getRequest(), getResponse(), getCallback()))
-                    Response.writeError(getRequest(), getResponse(), getCallback(), HttpStatus.NOT_FOUND_404);
+                if (getHandler().handle(request, response, callback))
+                    return true;
+
+                // The handle was rejected, so write the error using the original potentially unwrapped request/response/callback
+                Response.writeError(getRequest(), getResponse(), getCallback(), HttpStatus.NOT_FOUND_404);
             }
             catch (Throwable t)
             {
+                // The handle failed, so write the error using the original potentially unwrapped request/response/callback
                 Response.writeError(getRequest(), getResponse(), getCallback(), t);
             }
+            // return false to indicate the passed request/response/callback were not used.
+            return false;
         }
 
         protected abstract void delay() throws Exception;
     }
 
-    protected static class UntilContentDelayedProcess extends DelayedProcess
+    /**
+     * Delay dispatch until all content or an effective buffer size is reached
+     */
+    protected static class UntilContentDelayedProcess extends DelayedProcess implements Runnable
     {
-        public UntilContentDelayedProcess(Handler handler, Request request, Response response, Callback callback)
+        private final Deque<Content.Chunk> _chunks = new ArrayDeque<>();
+        private final int _maxSize;
+        private int _estimatedSize;
+
+        /**
+         * @param handler The next handler
+         * @param request The delayed request
+         * @param response The delayed response
+         * @param callback The delayed callback
+         * @param maxSize The maximum size to buffer before dispatching to the next handler;
+         *                or -1 to use {@link HttpConnectionFactory#getInputBufferSize()}
```

Comment:
Would not use `inputBufferSize`, but just the value, with `-1` indicating "until last chunk".

Reply:
inputbuffersize is a good default as it implies zero additional memory commitment. Any other value needs careful thought.

Reply:
Also the value `1` implies zero additional memory commitment.

The problem with the current implementation is that it tries to estimate the content length, and it gets it wrong for every protocol.

I would not try to complicate things.
By the time we read from the `Request`, it is actually content bytes without framing, so let's make `maxSize` indicating content bytes, not some wrongly estimated value.

My concern are that the current setup is weird and applications will be surprised because they configure N, but they read content up to a different value, and there is unnecessary additional computation to try to compensate for framing just to try to fit the "input buffer" model -- I don't understand why?

For HTTP/2 and HTTP/3 all the parsing has already been done, and the content is now sitting in the `Stream` objects. 
For HTTP/1, if there is data leftover in the buffer, it will be parsed by the application immediately.

Furthermore, if data flow control happens, trying to match the input buffer size would make applications wait unnecessarily, so `-1` would be a surprising default.

The previous semantic we had in Jetty 11 was "invoke at the first content byte", so I think our default should be that.

Reply:
No.  The wait until first byte can be really non optimal.   It does well handle the cases when the headers come in one packet and the bulk of the content comes in the next.   But it does not well handled the case when the content comes in multiple packets over time.
The default should be the best heuristic we can do, so for the most part, applications do not need to think about this configuration at all.    Your proposal is that in order to get the best optimization, then a developer has to think about it and pick a specific value.  Then we need to explain all the details and nuances of what is the right number to pick.   Much easier to just give a default that says we heuristically determine the best value and in most circumstances no further thought is needed.